src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -839,59 +854,194 @@ mergeRuns resolve hfs hbio caching alloc runPaths mergeLevel runs = do
+    -- Add credits to the credit accumulator, and try to do some merging work
+    -- if the accumulator now goes over the threshold.
+    prev <- fetchAddInt creditsAccumVar newCredits
+    let expectedNew = prev + newCredits
+
+    -- TODO: when a merge is already completed, then could we skip most of the
+    -- stuff below?
+    when (expectedNew >= creditsThresh) $ do
+
+      isMergeDone <-
+        -- Try to take credits from the accumulator. The number of taken credits
+        -- is the number of merging steps we do.
+        --
+        -- If an error happens during or after the merging work, then we put
+        -- back as many credits as we took, even if the merge has progressed. A
+        -- merge may subsequently finish earlier than expected, but that is
+        -- fine. If we did not put back credits, however, then a merge might not
+        -- finish in time, which is worse.
+        bracketOnError (takeCredits (Credit expectedNew)) putBackCredits $ \case
+          Nothing -> pure False
+          Just (Credit stepsToMerge) ->
+            withMVar mergeVar $ \case
+              CompletedMerge{} -> pure False
+              (OngoingMerge _rs numStepsDoneVar m) -> do
+                (n, stepResult) <- Merge.steps m stepsToMerge
+                -- If an asynchronous exception here, then the merge has
+                -- progressed more than is recorded in numStepsDoneVar. That is
+                -- okay, because the credits are put back.
+                void $ fetchAddInt numStepsDoneVar n
+                pure $ stepResult == MergeDone
+
+      -- If we just finished the merge, then we convert the output of the merge
+      -- into a new run. i.e., we complete the merge.
+      --
+      -- If an async exception happens before we get to perform the completion,
+      -- then that is fine. Either the next supplyMergeCredits or
+      -- expectCompletedMerge will complete the merge.
+      when isMergeDone $ completeMerge mergeVar
+  where
+    -- In a CAS-loop, subtract as many credits from the accumulator as required
+    -- to get it below the threshold again. If succesful, return Just that many
+    -- credits, or Nothing otherwise.
+    --
+    -- Nothing can be returned if the accumulator has already gone below the
+    -- threshold, which may happen if another thread is concurrently doing the
+    -- same loop on the same PrimVar.
+    takeCredits :: Credit -> m (Maybe Credit)
+    takeCredits (Credit prev)
+      | prev < creditsThresh = pure Nothing
+      | otherwise = do
+          -- numThresholds is guaranteed to be >= 1
+          let numThresholds = prev `div` creditsThresh
+          prev' <- casInt creditsAccumVar prev (numThresholds * creditsThresh)
+          if prev' == prev then
+            pure (Just (Credit (numThresholds * creditsThresh)))
+          else
+            takeCredits (Credit prev')
```

Comment:
Also: should this reduce the accumulated credits to 0 instead of getting it just under the threshold?

Reply:
It should not, I think. It makes batches of merge work relatively uniform if we do it like it is written above. I've elaborated on this with a comment in the code.